File: src/Butil/Bit.Butil/Scripts/dataTransfer.ts
var BitButil = (window as any).BitButil = (window as any).BitButil || {};

(function (butil: any) {
    interface DropEntry { element: any; handlers: { [type: string]: (e: DragEvent) => void }; }
    // draggable is remembered as it was found: an element that was already draggable (or explicitly
    // draggable="false") has to be left that way when the source is removed, not stripped bare.
    interface SourceEntry { element: any; handler: (e: DragEvent) => void; draggable: string | null; }

    const _drops: { [id: string]: DropEntry } = {};
    const _sources: { [id: string]: SourceEntry } = {};
    // Dropped files, kept alive past the event that delivered them. A DataTransfer's contents are
    // only readable during the event; the File objects themselves stay valid afterwards, so this is
    // what lets .NET read the bytes on its own schedule instead of inside the handler.
    const _files: { [id: string]: File } = {};

    butil.dataTransfer = {
        isSupported() { return typeof DataTransfer === 'function'; },

        // Wires up the whole drop target, including the part everyone forgets: a drop event only
        // fires on an element whose dragover handler called preventDefault. Doing it here means a
        // target that works, rather than one that silently never fires.
        listenForDrop(dotNetRef: any, id: string, element: any, dropEffect: string) {
            if (!element) return false;

            const handlers: { [type: string]: (e: DragEvent) => void } = {};

            const over = (e: DragEvent) => {
                e.preventDefault();
                // Setting dropEffect is what changes the cursor, and it has to be set on every
                // dragover - the browser resets it between events.
                if (e.dataTransfer) e.dataTransfer.dropEffect = dropEffect as any;
            };

            handlers['dragenter'] = over;
            handlers['dragover'] = over;

            handlers['drop'] = (e: DragEvent) => {
                e.preventDefault();

                const transfer = e.dataTransfer;
                const files: any[] = [];
                const items: any = {};

                if (transfer) {
                    for (const file of Array.from(transfer.files ?? [])) {
                        const fileId = butil.utils.randomUUID();
                        _files[fileId] = file;
                        files.push({
                            id: fileId,
                            name: file.name,
                            size: file.size,
                            type: file.type,
                            // Named for what it is - the .NET side exposes a DateTimeOffset built
                            // from it, and the two have to agree on the property name to bind.
                            lastModifiedMilliseconds: file.lastModified
                        });
                    }

                    // Everything that is not a file: text/plain, text/uri-list, text/html, and
                    // whatever custom type the drag source set. getData outside the drop event
                    // returns an empty string, so it is read here and handed over whole.
                    for (const type of Array.from(transfer.types ?? [])) {
                        if (type === 'Files') continue;
                        items[type] = transfer.getData(type);
                    }
                }

                butil.utils.dispatch(dotNetRef, 'InvokeDrop', id, files, items);
            };

            for (const type of Object.keys(handlers)) element.addEventListener(type, handlers[type]);
            _drops[id] = { element, handlers };
            return true;
        },

        removeDropListener(id: string) {
            const entry = _drops[id];
            if (!entry) return;
            delete _drops[id];
            for (const type of Object.keys(entry.handlers)) {
                entry.element.removeEventListener(type, entry.handlers[type]);
            }
        },

        // The drag source. The payload is configured up front rather than produced by a callback,
        // because dragstart's data has to be set synchronously and a round trip to .NET is not.
        configureSource(id: string, element: any, items: any, effectAllowed: string, dragImage: any, imageX: number, imageY: number) {
            if (!element) return false;

            const handler = (e: DragEvent) => {
                const transfer = e.dataTransfer;
                if (!transfer) return;

                transfer.effectAllowed = effectAllowed as any;
                for (const type of Object.keys(items ?? {})) transfer.setData(type, items[type]);

                // The drag image has to be an element that is rendered and visible at the moment of
                // the call - a display:none one produces no image at all, which is the usual reason
                // a custom drag image does not appear.
                if (dragImage) {
                    try { transfer.setDragImage(dragImage, imageX, imageY); } catch { /* not a valid image source */ }
                }
            };

            const draggable = element.getAttribute('draggable');
            element.setAttribute('draggable', 'true');
            element.addEventListener('dragstart', handler);
            _sources[id] = { element, handler, draggable };
            return true;
        },

        removeSource(id: string) {
            const entry = _sources[id];
            if (!entry) return;
            delete _sources[id];
            entry.element.removeEventListener('dragstart', entry.handler);
            if (entry.draggable === null) entry.element.removeAttribute('draggable');
            else entry.element.setAttribute('draggable', entry.draggable);
        },

        async readFile(fileId: string) {
            const file = _files[fileId];
            if (!file) return null;
            try { return new Uint8Array(await file.arrayBuffer()); } catch { return null; }
        },

        async readFileText(fileId: string) {
            const file = _files[fileId];
            if (!file) return null;
            try { return await file.text(); } catch { return null; }
        },

        // An object URL for a dropped file, for showing an image without reading its bytes into
        // .NET first. The caller owns it and has to revoke it.
        objectUrl(fileId: string) {
            const file = _files[fileId];
            return file ? URL.createObjectURL(file) : null;
        },

        releaseFile(fileId: string) { delete _files[fileId]; },

        disposeAll() {
            for (const id of Object.keys(_drops)) butil.dataTransfer.removeDropListener(id);
            for (const id of Object.keys(_sources)) butil.dataTransfer.removeSource(id);
            for (const id of Object.keys(_files)) delete _files[id];
        }
    };
}(BitButil));

File: src/Butil/Bit.Butil/Publics/DataTransfer.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Components;
using Microsoft.JSInterop;

namespace Bit.Butil;

/// <summary>
/// Wraps <see href="https://developer.mozilla.org/en-US/docs/Web/API/DataTransfer">DataTransfer</see>:
/// what a drag is actually carrying - the files, the text, the effect and the drag image.
/// </summary>
/// <remarks>
/// Blazor's <c>DragEventArgs</c> gives you the event but not the payload: <c>files</c>,
/// <c>getData</c>/<c>setData</c>, <c>dropEffect</c> and <c>setDragImage</c> are all out of reach, so
/// handling a file dropped onto the page has meant hand-written JavaScript. This is that JavaScript.
/// <br/>
/// Two rules of the drag-and-drop API are handled for you, because getting either wrong produces a
/// target that silently never fires: a <c>drop</c> only happens on an element whose <c>dragover</c>
/// handler called <c>preventDefault</c>, and <c>dropEffect</c> has to be set on <em>every</em>
/// dragover because the browser resets it between events.
/// <br/>
/// A dropped file's contents are readable long after the event that delivered it - the file objects
/// are held for you, and <see cref="ReadFile"/> reads them on your own schedule rather than inside
/// the handler. Release them when you are done.
/// </remarks>
[ButilService(typeof(DataTransfer))]
public class DataTransfer(IJSRuntime js) : IAsyncDisposable
{
    internal const string DropMethodName = nameof(InvokeDrop);

    private readonly ConcurrentDictionary<Guid, Action<DropPayload>> _dropHandlers = new();

    // Per-instance callback reference (see Keyboard): drop targets are isolated per circuit / WASM
    // app and detached on disposal - no static state, no cross-circuit leak.
    private DotNetObjectReference<DataTransfer>? _dotNetRef;
    private DotNetObjectReference<DataTransfer> DotNetRef => DotNetObjectReferenceHelper.GetOrCreate(ref _dotNetRef, this);

    /// <summary>True when the runtime exposes <c>DataTransfer</c>, which is everywhere.</summary>
    /// <remarks>
    /// During prerender/SSR (no JS runtime) this returns <c>default</c> (<c>false</c>) rather than
    /// throwing, so the result can't be distinguished from a genuine value.
    /// </remarks>
    public ValueTask<bool> IsSupported() => js.Invoke<bool>("BitButil.dataTransfer.isSupported");

    /// <summary>
    /// Invoked from JS when something is dropped. Public + <see cref="JSInvokableAttribute"/> so it
    /// can be dispatched through the per-instance <see cref="DotNetObjectReference{T}"/>.
    /// </summary>
    [JSInvokable(DropMethodName)]
    public void InvokeDrop(Guid id, DroppedFile[] files, Dictionary<string, string> items)
    {
        if (_dropHandlers.TryGetValue(id, out var handler)) handler(new DropPayload(files ?? [], items ?? []));
    }

    /// <summary>
    /// Makes an element a drop target and reports what lands on it.
    /// </summary>
    /// <param name="target">The element to accept drops on.</param>
    /// <param name="onDrop">
    /// Called with the payload: the files, and every non-file item the drag was carrying keyed by
    /// its MIME type - <c>"text/plain"</c>, <c>"text/uri-list"</c>, <c>"text/html"</c>, or whatever
    /// custom type the source set.
    /// </param>
    /// <param name="dropEffect">
    /// The cursor to show while a drag is over the target: <c>"copy"</c> (the default),
    /// <c>"move"</c>, <c>"link"</c> or <c>"none"</c>. It says what <em>will</em> happen; doing it is
    /// still yours.
    /// </param>
    /// <returns>
    /// A subscription; disposing it stops the element being a drop target. Null when the target
    /// could not be wired up - there is no element behind the reference, or there is no JS runtime
    /// to reach it with (prerender/SSR).
    /// </returns>
    /// <remarks>
    /// The <c>dragover</c> handling that makes a drop possible at all is wired up with this, so
    /// there is nothing else to add.
    /// </remarks>
    [DynamicDependency(nameof(InvokeDrop), typeof(DataTransfer))]
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(DroppedFile))]
    public async ValueTask<ButilSubscription?> OnDrop(ElementReference target, Action<DropPayload> onDrop, string dropEffect = "copy")
    {
        ArgumentNullException.ThrowIfNull(onDrop);

        var id = Guid.NewGuid();
        _dropHandlers[id] = onDrop;

        var listening = await js.Invoke<bool>("BitButil.dataTransfer.listenForDrop", DotNetRef, id, target, dropEffect);

        if (listening is false)
        {
            // Nothing is listening, so a subscription would be a handle on nothing - and the
            // handler it kept would never be called and never be removed.
            _dropHandlers.TryRemove(id, out _);
            return null;
        }

        return new ButilSubscription(id, async () =>
        {
            _dropHandlers.TryRemove(id, out _);
            await js.InvokeVoid("BitButil.dataTransfer.removeDropListener", id);
        });
    }

    /// <summary>
    /// Makes an element draggable and decides what the drag carries.
    /// </summary>
    /// <param name="source">The element to drag.</param>
    /// <param name="items">
    /// What to carry, keyed by MIME type. <c>"text/plain"</c> is what other applications will read;
    /// a custom type like <c>"application/x-my-app"</c> is how a drop target of your own recognises
    /// its own drags.
    /// </param>
    /// <param name="effectAllowed">
    /// What the source permits: <c>"copy"</c>, <c>"move"</c>, <c>"link"</c>, <c>"copyMove"</c>,
    /// <c>"all"</c> (the default), <c>"none"</c>. A target asking for an effect this does not allow
    /// gets no drop.
    /// </param>
    /// <param name="dragImage">
    /// An element to show under the cursor instead of a snapshot of the source. It has to be
    /// rendered and visible when the drag starts - a <c>display:none</c> one produces no image at
    /// all, which is the usual reason a custom drag image does not appear.
    /// </param>
    /// <param name="dragImageX">Where the cursor sits within the image, horizontally.</param>
    /// <param name="dragImageY">Where the cursor sits within the image, vertically.</param>
    /// <returns>
    /// A subscription; disposing it makes the element undraggable again. Null when the source could
    /// not be wired up - there is no element behind the reference, or there is no JS runtime to
    /// reach it with (prerender/SSR).
    /// </returns>
    /// <remarks>
    /// The payload is settled here rather than produced by a callback because <c>dragstart</c> has
    /// to set its data synchronously, and a round trip to .NET is not synchronous. Call this again
    /// (after disposing) when what the element carries changes.
    /// </remarks>
    public async ValueTask<ButilSubscription?> ConfigureDragSource(ElementReference source,
                                                                   Dictionary<string, string> items,
                                                                   string effectAllowed = "all",
                                                                   ElementReference? dragImage = null,
                                                                   int dragImageX = 0,
                                                                   int dragImageY = 0)
    {
        ArgumentNullException.ThrowIfNull(items);

        var id = Guid.NewGuid();
        var configured = await js.Invoke<bool>("BitButil.dataTransfer.configureSource",
            id, source, items, effectAllowed, dragImage, dragImageX, dragImageY);

        if (configured is false) return null;

        return new ButilSubscription(id, async () => await js.InvokeVoid("BitButil.dataTransfer.removeSource", id));
    }

    /// <summary>
    /// Reads a dropped file's bytes.
    /// </summary>
    /// <returns>The bytes, or null when the file has been released or could not be read.</returns>
    /// <remarks>
    /// Readable long after the drop - the file object is held for you, so there is no need to do
    /// the work inside the handler. Release it when you are done.
    /// </remarks>
    public ValueTask<byte[]?> ReadFile(Guid fileId) => js.Invoke<byte[]?>("BitButil.dataTransfer.readFile", fileId);

    /// <summary>
    /// Reads a dropped file as UTF-8 text.
    /// </summary>
    /// <returns>The text, or null when the file has been released or could not be read.</returns>
    public ValueTask<string?> ReadFileText(Guid fileId) => js.Invoke<string?>("BitButil.dataTransfer.readFileText", fileId);

    /// <summary>
    /// A <c>blob:</c> URL for a dropped file - for putting a dropped image straight into an
    /// <c>&lt;img src&gt;</c> without reading its bytes into .NET at all.
    /// </summary>
    /// <returns>The URL, or null when the file has been released.</returns>
    /// <remarks>
    /// You own the URL: revoke it through <see cref="ObjectUrls"/> when the image is gone, or the
    /// file stays in memory for the life of the page.
    /// </remarks>
    public ValueTask<string?> CreateObjectUrl(Guid fileId) => js.Invoke<string?>("BitButil.dataTransfer.objectUrl", fileId);

    /// <summary>
    /// Forgets a dropped file. Reading it afterwards answers null.
    /// </summary>
    public ValueTask ReleaseFile(Guid fileId) => js.InvokeVoid("BitButil.dataTransfer.releaseFile", fileId);

    /// <summary>
    /// On scope/circuit teardown, detaches every drop target and drag source whose subscription was
    /// never disposed, and forgets every dropped file still being held.
    /// </summary>
    public async ValueTask DisposeAsync()
    {
        try
        {
            _dropHandlers.Clear();
            await js.InvokeVoid("BitButil.dataTransfer.disposeAll");
        }
        catch (Exception ex) when (ex.IsIgnorableDisposalException()) { } // teardown: circuit gone, cancelled, or already disposed
        finally
        {
            _dotNetRef?.Dispose();
            _dotNetRef = null;
        }
        GC.SuppressFinalize(this);
    }
}
